### ISSUE TITLE
Valid query throws ArgumentException

### ISSUE THREAD

**Issue by JoasE** [aspnet/EntityFrameworkCore#14991]
A valid LINQ query which works over a `List` throws an `System.ArgumentException` when querying over a `DbSet`.

The expression is copied from an expression generated by AutoMapper, but the expression should be valid.

Moved from: AutoMapper/AutoMapper#3010

```
Exception message: System.ArgumentException: `Argument types do not match`
Stack trace: at System.Linq.Expressions.Expression.Bind(MemberInfo member, Expression expression)
   at System.Linq.Expressions.MemberAssignment.Update(Expression expression)
   at System.Linq.Expressions.ExpressionVisitor.VisitMemberAssignment(MemberAssignment node)
   at System.Linq.Expressions.ExpressionVisitor.VisitMemberBinding(MemberBinding node)
   at System.Linq.Expressions.ExpressionVisitor.Visit[T](ReadOnlyCollection`1 nodes, Func`2 elementVisitor)
   at System.Linq.Expressions.ExpressionVisitor.VisitMemberInit(MemberInitExpression node)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.RelationalProjectionExpressionVisitor.VisitMemberInit(MemberInitExpression memberInitExpression)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.RelationalProjectionExpressionVisitor.Visit(Expression expression)
   at System.Linq.Expressions.ExpressionVisitor.VisitConditional(ConditionalExpression node)
   at System.Linq.Expressions.ConditionalExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.RelationalProjectionExpressionVisitor.Visit(Expression expression)
   at System.Linq.Expressions.ExpressionVisitor.VisitMemberAssignment(MemberAssignment node)
   at System.Linq.Expressions.ExpressionVisitor.VisitMemberBinding(MemberBinding node)
   at System.Linq.Expressions.ExpressionVisitor.Visit[T](ReadOnlyCollection`1 nodes, Func`2 elementVisitor)
   at System.Linq.Expressions.ExpressionVisitor.VisitMemberInit(MemberInitExpression node)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.RelationalProjectionExpressionVisitor.VisitMemberInit(MemberInitExpression memberInitExpression)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.RelationalProjectionExpressionVisitor.Visit(Expression expression)
   at System.Linq.Expressions.ExpressionVisitor.VisitAndConvert[T](ReadOnlyCollection`1 nodes, String callerName)
   at Remotion.Linq.Parsing.RelinqExpressionVisitor.VisitNew(NewExpression expression)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.RelationalProjectionExpressionVisitor.VisitNew(NewExpression newExpression)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.RelationalProjectionExpressionVisitor.Visit(Expression expression)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.VisitSelectClause(SelectClause selectClause, QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.VisitSelectClause(SelectClause selectClause, QueryModel queryModel)
   at Remotion.Linq.Clauses.SelectClause.Accept(IQueryModelVisitor visitor, QueryModel queryModel)
   at Remotion.Linq.QueryModelVisitorBase.VisitQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.VisitQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.VisitQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.RelationalEntityQueryableExpressionVisitor.VisitSubQuery(SubQueryExpression expression)
   at Remotion.Linq.Clauses.Expressions.SubQueryExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.ExpressionVisitorBase.VisitLambda[T](Expression`1 node)
   at System.Linq.Expressions.Expression`1.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at System.Linq.Expressions.ExpressionVisitor.VisitArguments(IArgumentProvider nodes)
   at System.Linq.Expressions.ExpressionVisitor.VisitMethodCall(MethodCallExpression node)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.RelationalEntityQueryableExpressionVisitor.VisitMethodCall(MethodCallExpression node)
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at System.Linq.Expressions.ExpressionVisitor.VisitMemberAssignment(MemberAssignment node)
   at System.Linq.Expressions.ExpressionVisitor.VisitMemberBinding(MemberBinding node)
   at System.Linq.Expressions.ExpressionVisitor.Visit[T](ReadOnlyCollection`1 nodes, Func`2 elementVisitor)
   at System.Linq.Expressions.ExpressionVisitor.VisitMemberInit(MemberInitExpression node)
   at System.Linq.Expressions.MemberInitExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.ReplaceClauseReferences(Expression expression, IQuerySource querySource, Boolean inProjection)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.VisitSelectClause(SelectClause selectClause, QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.VisitSelectClause(SelectClause selectClause, QueryModel queryModel)
   at Remotion.Linq.Clauses.SelectClause.Accept(IQueryModelVisitor visitor, QueryModel queryModel)
   at Remotion.Linq.QueryModelVisitorBase.VisitQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.VisitQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.VisitQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.CreateQueryExecutor[TResult](QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery[TResult](QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore[TResult](Expression query, IQueryModelGenerator queryModelGenerator, IDatabase database, IDiagnosticsLogger`1 logger, Type contextType)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass13_0`1.<Execute>b__0()
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQueryCore[TFunc](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.Execute[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.Execute[TResult](Expression expression)
   at System.Linq.Queryable.FirstOrDefault[TSource](IQueryable`1 source, Expression`1 predicate)
   at ConsoleApp5.Program.Main(String[] args) in C:\Users\Joas\source\repos\ConsoleApp5\ConsoleApp5\Program.cs:line 47
```

### Steps to reproduce

https://github.com/JoasE/AutoMapperRepro

```C#
var manualQuery = context.Users.Select(dtoUser => new User.ViewDto
{
	Id = dtoUser.Id,
	Posts = dtoUser.Posts
		.Select(
			dtoPost => new Object_205617611___LastComment_Id_Text_UserId
			{
				__LastComment = dtoPost.Comments.LastOrDefault(),
				Id = dtoPost.Id,
				Text = dtoPost.Text,
				UserId = dtoPost.UserId
			})
		.Select(
			dtoLet => new Post.ViewDto
			{
				Id = dtoLet.Id,
				LastComment = (dtoLet.__LastComment == null)
					? null
					: new Comment.ViewDto
					{
						Id = dtoLet.__LastComment.Id,
						PostId = dtoLet.__LastComment.PostId,
						Text = dtoLet.__LastComment.Text
					},
				Text = dtoLet.Text,
				UserId = dtoLet.UserId
			})
		.ToList()
}).FirstOrDefault(x => x.Id == user.Id);
```

### Further technical details
EF Core version: 2.2.2
Database Provider: Microsoft.EntityFrameworkCore.SqlServer 2.2.2
Operating system: Windows 10
IDE: Visual Studio 2017 15.9.6

**Comment by gsGabriel** [comment]
I have the same issue... But with the Npgsql.EntityFrameworkCore.PostgreSQL 2.2.0

**Comment by JoasE** [comment]
@gsGabriel Are you sure it's the same issue? Is your query and data structure similar to mine? Could you share your reproduce steps? It might help the guys debug and fix this issue.

**Comment by gsGabriel** [comment]
Yes, I am sure, the data structure is not exactly the same but it is very close... Follows the sample that resembles my
```
var manualQuery = await context.Users.Include(x => x.UsersPosts).ThenInclude(x => x.Posts)
.Select(dtoUser => new User.ViewDto
{
	Posts = dtoUser.UsersPosts.Posts.Select(dtoPost  => new Post.ViewDto { dtoPost.Id })
}).ToArrayAsync();

```

**Comment by gsGabriel** [comment]
To solve, i broke between two queries, the first to get the posts, on the second query i`m used the first query to binding a dto.

**Comment by JoasE** [comment]
I've tested this and can confirm this has been fixed as of EntityFramework 3.0.0, **with some changes made to the query.**

When I ran the original code in 3.0.0 I would get an exception:
```
System.InvalidOperationException: 'The LINQ expression 'LastOrDefault<Nullable<Guid>>(Select<Comment, Nullable<Guid>>(
    source: Where<Comment>(
        source: DbSet<Comment>, 
        predicate: (c) => Property<Nullable<Guid>>(EntityShaperExpression: 
            EntityType: Post
            ValueBufferExpression: 
                ProjectionBindingExpression: EmptyProjectionMember
            IsNullable: False
        , "Id") == Property<Nullable<Guid>>(c, "PostId")), 
    selector: (c) => Property<Nullable<Guid>>(c, "Id")))' could not be translated. Either rewrite the query in a form that can be translated, or switch to client evaluation explicitly by inserting a call to either AsEnumerable(), AsAsyncEnumerable(), ToList(), or ToListAsync(). See https://go.microsoft.com/fwlink/?linkid=2101038 for more information.'
```

After a quick google search I found [this issue](https://github.com/aspnet/EntityFrameworkCore/issues/10493), so I realized I could change the query in order for it to work:

```csharp
var manualQuery = context.Users.Select(dtoUser => new User.ViewDto
{
    Id = dtoUser.Id,
    Posts = dtoUser.Posts
        .Select(
            dtoPost => new Object_205617611___LastComment_Id_Text_UserId
            {
                __LastComment = dtoPost.Comments.OrderBy(x => x.CreatedAt).LastOrDefault(), // I have added the CreatedAt property
                Id = dtoPost.Id,
                Text = dtoPost.Text,
                UserId = dtoPost.UserId
            })
        .Select(
            dtoLet => new Post.ViewDto
            {
                Id = dtoLet.Id,
                LastComment = (dtoLet.__LastComment == null)
                    ? null
                    : new Comment.ViewDto
                    {
                        Id = dtoLet.__LastComment.Id,
                        PostId = dtoLet.__LastComment.PostId,
                        Text = dtoLet.__LastComment.Text
                    },
                Text = dtoLet.Text,
                UserId = dtoLet.UserId
            })
        .ToList()
}).FirstOrDefault(x => x.Id == user.Id);
```

or

```csharp
__LastComment = dtoPost.Comments.OrderByDescending(x => x.CreatedAt).FirstOrDefault(), 
````

My thanks to the team and any contributors who might have helped for fixing this issue.
